tests/worker/test_workflow.py

```diff
@@ -0,0 +1,989 @@
+
+
+async def test_workflow_info(client: Client):
+    async with new_worker(client, InfoWorkflow) as worker:
+        workflow_id = f"workflow-{uuid.uuid4()}"
+        info = await client.execute_workflow(
+            InfoWorkflow.run, id=workflow_id, task_queue=worker.task_queue
+        )
+        assert info["attempt"] == 1
+        assert info["cron_schedule"] is None
+        assert info["execution_timeout"] is None
+        assert info["namespace"] == client.namespace
+        assert uuid.UUID(info["run_id"]).version == 4
+        assert info["run_timeout"] is None
+        datetime.fromisoformat(info["start_time"])
+        assert info["task_queue"] == worker.task_queue
+        assert info["task_timeout"] == "0:00:10"
+        assert info["workflow_id"] == workflow_id
+        assert info["workflow_type"] == "InfoWorkflow"
+
+
+@workflow.defn
+class SignalAndQueryWorkflow:
+    def __init__(self) -> None:
+        self._last_event: Optional[str] = None
+
+    @workflow.run
+    async def run(self) -> None:
+        # Wait forever
+        await asyncio.Future()
+
+    @workflow.signal
+    def signal1(self, arg: str) -> None:
+        self._last_event = f"signal1: {arg}"
+
+    @workflow.signal(dynamic=True)
+    def signal_dynamic(self, name: str, *args: Any) -> None:
+        self._last_event = f"signal_dynamic {name}: {args[0]}"
+
+    @workflow.signal(name="Custom Name")
+    def signal_custom(self, arg: str) -> None:
+        self._last_event = f"signal_custom: {arg}"
+
+    @workflow.query
+    def last_event(self) -> str:
+        return self._last_event or "<no event>"
+
+    @workflow.query(dynamic=True)
+    def query_dynamic(self, name: str, *args: Any) -> str:
+        return f"query_dynamic {name}: {args[0]}"
+
+    @workflow.query(name="Custom Name")
+    def query_custom(self, arg: str) -> str:
+        return f"query_custom: {arg}"
+
+
+async def test_workflow_signal_and_query(client: Client):
+    async with new_worker(client, SignalAndQueryWorkflow) as worker:
+        handle = await client.start_workflow(
+            SignalAndQueryWorkflow.run,
+            id=f"workflow-{uuid.uuid4()}",
+            task_queue=worker.task_queue,
+        )
+
+        async def last_event() -> str:
+            return await handle.query(SignalAndQueryWorkflow.last_event)
+
+        # Simple signals and queries
+        await handle.signal(SignalAndQueryWorkflow.signal1, "some arg")
+        await assert_eq_eventually("signal1: some arg", last_event)
+
+        # Dynamic signals and queries
+        await handle.signal("signal2", "dyn arg")
+        await assert_eq_eventually("signal_dynamic signal2: dyn arg", last_event)
+        assert "query_dynamic query2: dyn arg" == await handle.query(
+            "query2", "dyn arg"
+        )
+
+        # Custom named signals and queries
+        await handle.signal("Custom Name", "custom arg1")
+        await assert_eq_eventually("signal_custom: custom arg1", last_event)
```

Comment:
This should be available immediately with consistent query

Reply:
True. Signal-then-query should always mean signal processed first. I will look into updating the tests to check this immediately in these cases.

Reply:
Fixed.